### PR TITLE
Add ExecuteAsync, Fix CancelAsync Deadlock

### PR DESCRIPTION
**Pull request by zeotuan**
- Implemented TAP over APM ExecuteAsync(CancellationToken) mentiond in https://github.com/sshnet/SSH.NET/issues/931
- Fix CancelAsync causing deadlocking and it now send SIGTERM signal to try to abort the request which will probably also solve https://github.com/sshnet/SSH.NET/issues/1147 and https://github.com/sshnet/SSH.NET/issues/1023

### CODE REVIEW

File: src/Renci.SshNet/SshCommand.cs [hunk 1]
```diff
@@ -348,21 +367,109 @@ public string EndExecute(IAsyncResult asyncResult)
...
+        {
+#if NET || NETSTANDARD2_1_OR_GREATER
+            await using var ctr = cancellationToken.Register(() => CancelAsync(forceKill), useSynchronizationContext: false).ConfigureAwait(continueOnCapturedContext: false);
+#else
+            using var ctr = cancellationToken.Register(() => CancelAsync(forceKill), useSynchronizationContext: false);
+#endif // NET || NETSTANDARD2_1_OR_GREATER
+
+            try
+            {
+                var status = await Task<int>.Factory.FromAsync(BeginExecute(), EndExecuteWithStatus).ConfigureAwait(false);
```
Comment (WojciechNagorski) [inline src/Renci.SshNet/SshCommand.cs:453]: I want to add `Async` overloaded methods but I think this is not a real `async` code. I've prepared a POC here: https://github.com/sshnet/SSH.NET/pull/1318 

I know that your approach is easier to implement, but this is similar to nuget: https://www.nuget.org/packages/Renci.SshNet.Async#versions-body-tab

You can find code here: https://github.com/JohnTheGr8/Renci.SshNet.Async/blob/master/Renci.SshNet.Async/SshNetExtensions.cs

File: src/Renci.SshNet/SshCommand.cs [hunk 2]
```diff
@@ -105,56 +107,72 @@ public Stream CreateInputStream()
         /// <summary>
         /// Gets the command execution result.
         /// </summary>
+#pragma warning disable S1133 // Deprecated code should be removed
+        [Obsolete("Please read the result from the OutputStream. I.e. new StreamReader(shell.OutputStream).ReadToEnd().")]
```
Comment (WojciechNagorski) [inline src/Renci.SshNet/SshCommand.cs:111]: Why should we make it obsolete?

File: src/Renci.SshNet/SshCommand.cs [hunk 3]
```diff
@@ -105,56 +107,72 @@ public Stream CreateInputStream()
...
             }
+
+            return _result.ToString();
         }
 
         /// <summary>
         /// Gets the command execution error.
         /// </summary>
+#pragma warning disable S1133 // Deprecated code should be removed
+        [Obsolete("Please read the error result from the ExtendedOutputStream. I.e. new StreamReader(shell.ExtendedOutputStream).ReadToEnd().")]
```
Comment (WojciechNagorski) [inline src/Renci.SshNet/SshCommand.cs:144]: Why should we make it obsolete?

**Comment by Rob-Hague** [comment]
@WojciechNagorski commented just as I was commenting some similar feedback, so sorry for the duplication 🙂 :

I think you are still working on this so just some high-level thoughts:

There are a few semi-related changes here. The cancellation via signal is good. I would suggest splitting that to a standalone PR to start with.

I understand the reason for obsoleting the Result/Error properties (because they consume the stream properties) but I do not think the change is justified. Especially given how many changes it causes just in our own code (forwarding to a now internal version at that). I would suggest reverting that.

I'm a little uneasy about using `TaskFactory.FromAsync` in this library because many of the Begin/End (APM) methods are not implemented so well and we end up spawning new threads/blocking on more threads than we need. I basically share the same sentiment as https://github.com/sshnet/SSH.NET/issues/931#issuecomment-1066160452. But at first glance it is not *so* bad here.

**Comment by zeotuan** [comment]
@Rob-Hague @WojciechNagorski Very Interesting. I would love to help with the project.

https://www.nuget.org/packages/Renci.SshNet.Async#versions-body-tab unfortunately does not support ```CancellationToken```. 
So I was thinking of adding a simple TAP over APM wrapper here until we get a proper TAP rewrite. 

I also expect the execution of command to leave the ```OutputStream``` and ```ExtendedOutputStream``` intact and provide my own way to consuming them. for example: I want to forward the stdout and stderr to another program while the command is being executed not after.

Therefore, I would suggest Future ExecuteAsync Implementation to at least just return an exit status code instead of consuming those stream.  Also Providing a public helper method to let people get Result/Error from those stream. It would require refactoring for any downstream project that will migrate to Async anyway

I have submitted a standalone PR for CancelAsync: https://github.com/sshnet/SSH.NET/pull/1345

**Comment by WojciechNagorski** [comment]
@zeotuan Sory for the delay. If you want you can continue to help us add true asynchronous methods. let us know.

**Comment by zeotuan** [comment]
@WojciechNagorski Hi! Yes, I would like to help with adding true async methods for sshnet.

**Pull request closed without merging**